File: apps/web/lib/api/pots.ts
/*
 * Server-only types + mappers for pots — the container co-ownership attaches to. Reads go through
 * the four functions here; mutations are server actions in `app/(protected)/shared/pot-actions.ts`.
 *
 * Two shapes of the API's honesty survive into these types unchanged, and the UI must respect both:
 *
 *   * `nav` and `unitPrice` are `null` when unknown, never 0. A pot with no holdings, or one holding
 *     a currency with no stored rate, has no value to state — and "we do not know" must never render
 *     as "it is worth nothing", because only one of those is safe to price units against.
 *   * A pot the caller may not see answers 404 exactly as one that does not exist, so `getPot`
 *     returns null for both and the page renders notFound() either way. An id cannot be probed.
 *
 * Money and unit figures stay STRINGS end to end, as everywhere else in the app: they are
 * NUMERIC(18,2) / NUMERIC(18,6) server-side, and parsing them into a JS number would round exactly
 * the digits the ledger exists to keep.
 */

import 'server-only';

import { authenticatedFetch } from '@/lib/authenticated-fetch';
import type {
  OwnershipEventType,
  PotCadence,
  PotSeriesInterval,
  PotVisibility,
} from '@/lib/constants/pots';

// --- Raw types (API JSON shape, snake_case) ---

interface PotMemberShareRaw {
  member_id: number;
  display_name: string;
  units: string;
  percentage: string;
  value: string | null;
  is_self: boolean;
}

interface PotPermissionRaw {
  member_id: number;
  display_name: string;
  can_view: boolean;
  can_write: boolean;
}

export interface PotRaw {
  id: number;
  group_id: number;
  name: string | null;
  base_currency: string;
  snapshot_cadence: PotCadence;
  visibility: PotVisibility;
  is_default: boolean;
  nav: string | null;
  valued_as_of: string | null;
  is_stale: boolean;
  unit_price: string | null;
  total_units: string;
  my_percentage: string;
  can_write: boolean;
  shares: PotMemberShareRaw[];
  permissions: PotPermissionRaw[];
  created_at: string;
  updated_at: string;
}

interface PotHoldingRaw {
  id: number;
  name: string;
  currency: string;
  value: string | null;
  base_value: string | null;
  is_active: boolean;
  valued_on: string | null;
}

interface PotHoldingsRaw {
  investments: PotHoldingRaw[];
  accounts: PotHoldingRaw[];
}

interface PotValueSeriesPointRaw {
  date: string;
  nav: string | null;
  my_value: string | null;
}

interface PotValueSeriesRaw {
  interval: PotSeriesInterval;
  points: PotValueSeriesPointRaw[];
}

interface PotOwnershipEventRaw {
  id: number;
  pot_id: number;
  type: OwnershipEventType;
  date: string;
  member_id: number;
  member_name: string;
  counterparty_member_id: number | null;
  counterparty_name: string | null;
  amount: string | null;
  amount_currency: string | null;
  base_amount: string | null;
  units: string;
  unit_price: string;
  from_account_id: number | null;
  to_account_id: number | null;
  notes: string | null;
  created_at: string;
}

// --- Frontend types (camelCase) ---

export interface PotMemberShare {
  memberId: number;
  displayName: string;
  units: string;
  percentage: string;
  // Share value in the pot's base currency; null when the NAV is unknown.
  value: string | null;
  isSelf: boolean;
}

export interface PotPermission {
  memberId: number;
  displayName: string;
  canView: boolean;
  canWrite: boolean;
}

export interface Pot {
  id: number;
  groupId: number;
  // Null for a group's default pot: the container is deliberately not a thing to name until there
  // is a second one to tell apart.
  name: string | null;
  baseCurrency: string;
  // How often the pot is expected to be re-valued. An expectation, not a schedule — it decides what
  // counts as an overdue valuation and how far apart the value series' points sit, nothing else.
  snapshotCadence: PotCadence;
  visibility: PotVisibility;
  isDefault: boolean;
  nav: string | null;
  /*
   * The date the value is current to, and whether the cadence considers it overdue. A different
   * question from `nav`, and the three can honestly disagree: a pot whose holdings are all
   * snapshotted but one of which cannot be converted has a real `valuedAsOf` and a null `nav`.
   * `valuedAsOf` is null when nothing here has ever been valued, which is a state of its own — the
   * page says "never valued" rather than showing a date.
   */
  valuedAsOf: string | null;
  isStale: boolean;
  unitPrice: string | null;
  totalUnits: string;
  myPercentage: string;
  canWrite: boolean;
  // Largest holder first, and only members actually holding units — someone bought out entirely is
  // not a 0.00% row on every screen.
  shares: PotMemberShare[];
  permissions: PotPermission[];
  createdAt: string;
  updatedAt: string;
}

export interface PotHolding {
  id: number;
  name: string;
  currency: string;
  value: string | null;
  baseValue: string | null;
  isActive: boolean;
  // The date this figure was recorded on. Null for an ACCOUNT, whose balance is derived at the date
  // it is asked for rather than recorded on one — so there is no valuation date to state.
  valuedOn: string | null;
}

export interface PotHoldings {
  investments: PotHolding[];
  accounts: PotHolding[];
}

export interface PotValueSeriesPoint {
  date: string;
  // Both null in the same case and for the same reason the NAV is. `myValue` is additionally null
  // while no units are outstanding: before the baseline nobody owns any share of anything.
  nav: string | null;
  myValue: string | null;
}

export interface PotValueSeries {
  // The grid the points sit on, which an `ad_hoc` pot's cadence does not name — it is plotted monthly.
  interval: PotSeriesInterval;
  points: PotValueSeriesPoint[];
}

export interface PotOwnershipEvent {
  id: number;
  potId: number;
  type: OwnershipEventType;
  date: string;
  memberId: number;
  memberName: string;
  counterpartyMemberId: number | null;
  counterpartyName: string | null;
  // Money moved in the PRIVATE account's currency; baseAmount is the same movement in the pot's.
  amount: string | null;
  amountCurrency: string | null;
  baseAmount: string | null;
  units: string;
  unitPrice: string;
  fromAccountId: number | null;
  toAccountId: number | null;
  notes: string | null;
  createdAt: string;
}

// --- Mappers ---

function mapShare(raw: PotMemberShareRaw): PotMemberShare {
  return {
    memberId: raw.member_id,
    displayName: raw.display_name,
    units: raw.units,
    percentage: raw.percentage,
    value: raw.value,
    isSelf: raw.is_self,
  };
}

function mapPermission(raw: PotPermissionRaw): PotPermission {
  return {
    memberId: raw.member_id,
    displayName: raw.display_name,
    canView: raw.can_view,
    canWrite: raw.can_write,
  };
}

export function mapPot(raw: PotRaw): Pot {
  return {
    id: raw.id,
    groupId: raw.group_id,
    name: raw.name,
    baseCurrency: raw.base_currency,
    snapshotCadence: raw.snapshot_cadence,
    visibility: raw.visibility,
    isDefault: raw.is_default,
    nav: raw.nav,
    valuedAsOf: raw.valued_as_of,
    isStale: raw.is_stale,
    unitPrice: raw.unit_price,
    totalUnits: raw.total_units,
    myPercentage: raw.my_percentage,
    canWrite: raw.can_write,
    shares: raw.shares.map(mapShare),
    permissions: raw.permissions.map(mapPermission),
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

function mapHolding(raw: PotHoldingRaw): PotHolding {
  return {
    id: raw.id,
    name: raw.name,
    currency: raw.currency,
    value: raw.value,
    baseValue: raw.base_value,
    isActive: raw.is_active,
    valuedOn: raw.valued_on,
  };
}

function mapValueSeries(raw: PotValueSeriesRaw): PotValueSeries {
  return {
    interval: raw.interval,
    points: raw.points.map((point) => ({
      date: point.date,
      nav: point.nav,
      myValue: point.my_value,
    })),
  };
}

function mapOwnershipEvent(raw: PotOwnershipEventRaw): PotOwnershipEvent {
  return {
    id: raw.id,
    potId: raw.pot_id,
    type: raw.type,
    date: raw.date,
    memberId: raw.member_id,
    memberName: raw.member_name,
    counterpartyMemberId: raw.counterparty_member_id,
    counterpartyName: raw.counterparty_name,
    amount: raw.amount,
    amountCurrency: raw.amount_currency,
    baseAmount: raw.base_amount,
    units: raw.units,
    unitPrice: raw.unit_price,
    fromAccountId: raw.from_account_id,
    toAccountId: raw.to_account_id,
    notes: raw.notes,
    createdAt: raw.created_at,
  };
}

// --- API functions ---

// Every pot the caller may see, optionally narrowed to one group. A pot they own 0% of is included:
// membership is not ownership, and the monitoring surface is not gated on holding a share.
export async function getPots(groupId?: number): Promise<Pot[]> {
  const endpoint = groupId === undefined ? '/pots' : `/pots?group_id=${groupId}`;
  const res = await authenticatedFetch(endpoint, { method: 'GET' });
  if (!res.ok) throw new Error('Failed to fetch pots');
  const raw: PotRaw[] = await res.json();
  return raw.map(mapPot);
}

/*
 * One pot with its ownership breakdown. Returns null for a pot that does not exist OR that the caller
 * may not see — the API answers 404 for both, so the page renders notFound() either way and an id
 * cannot be used to discover which pots exist.
 *
 * `asOfDate` values it at a past date, which is what a back-dated guided flow needs: the share it is
 * about is worth what it was worth THEN, and pricing the confirmation step at today's figure would
 * state a number the event will not be recorded at.
 */
export async function getPot(potId: number, asOfDate?: string): Promise<Pot | null> {
  const query = asOfDate === undefined ? '' : `?as_of_date=${asOfDate}`;
  const res = await authenticatedFetch(`/pots/${potId}${query}`, { method: 'GET' });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error('Failed to fetch pot');
  return mapPot(await res.json());
}

// Everything the pot holds, archived holdings included and flagged. Null for the same two reasons
// getPot is, so a page can fetch both and take one notFound() decision.
export async function getPotHoldings(potId: number): Promise<PotHoldings | null> {
  const res = await authenticatedFetch(`/pots/${potId}/holdings`, { method: 'GET' });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error('Failed to fetch pot holdings');
  const raw: PotHoldingsRaw = await res.json();
  return { investments: raw.investments.map(mapHolding), accounts: raw.accounts.map(mapHolding) };
}

/*
 * The pot's value at each point of its cadence's grid, plus the caller's own share value at each.
 * Oldest first, the last point being today, and never starting before the pot's ownership anchor.
 *
 * Null for the same two reasons getPot is, so a page can fetch both and take one notFound() decision.
 * A point's figures are null wherever the value cannot be stated in full on that date — which on a
 * real pot is most of the early points, and drawing a zero there would read as growth from nothing.
 */
export async function getPotSeries(
  potId: number,
  periods?: number,
): Promise<PotValueSeries | null> {
  const query = periods === undefined ? '' : `?periods=${periods}`;
  const res = await authenticatedFetch(`/pots/${potId}/series${query}`, { method: 'GET' });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error('Failed to fetch pot value series');
  return mapValueSeries(await res.json());
}

// The pot's ownership ledger in replay order — oldest first, which is the order the unit balances are
// derived in and therefore the only order the history reads correctly in.
export async function getPotOwnershipEvents(potId: number): Promise<PotOwnershipEvent[] | null> {
  const res = await authenticatedFetch(`/pots/${potId}/ownership`, { method: 'GET' });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error('Failed to fetch pot ownership events');
  const raw: PotOwnershipEventRaw[] = await res.json();
  return raw.map(mapOwnershipEvent);
}

File: apps/api/app/routers/pots.py
from datetime import date as date_type

from fastapi import APIRouter, Query, status

from app.deps.auth import CurrentUser
from app.deps.db import AdminSessionDep, SessionDep
from app.schemas.pot import (
    PotCreate,
    PotHoldingsMove,
    PotHoldingsResponse,
    PotMovementCreate,
    PotOpeningCreate,
    PotOwnershipEventResponse,
    PotPermissionUpdate,
    PotReagreementCreate,
    PotResponse,
    PotUpdate,
)
from app.services import pot_ownership_service, pot_service

router = APIRouter(prefix="/pots", tags=["pots"])


# Lists every pot the user may see, each with its ownership breakdown. Optionally narrowed to one
# group. A pot the user owns 0% of is still listed — membership is not ownership.
@router.get("", response_model=list[PotResponse])
async def list_pots(
    current_user: CurrentUser,
    session: SessionDep,
    group_id: int | None = Query(default=None, description="Only pots in this group."),
) -> list[PotResponse]:
    return await pot_service.list_pots(session, current_user, group_id=group_id)


# Returns one pot with its ownership breakdown, optionally as at a past date. Returns 404 when it
# does not exist or the user may not see it — the same answer for both, so probing ids reveals
# nothing about pots that are merely hidden.
@router.get("/{pot_id}", response_model=PotResponse)
async def get_pot(
    pot_id: int,
    current_user: CurrentUser,
    session: SessionDep,
    as_of_date: date_type | None = Query(default=None, description="Date to value the pot at; defaults to today."),
) -> PotResponse:
    return await pot_service.get_pot(session, pot_id, current_user, as_of_date=as_of_date)


# Creates a pot in a group and seats its creator with full access. Group admin only (403 otherwise).
# Takes the privileged session because the pot's first permission row is the one its own RLS policy
# reads, so the insert cannot satisfy it — the same bootstrap group creation has.
@router.post("", response_model=PotResponse, status_code=status.HTTP_201_CREATED)
async def create_pot(
    body: PotCreate,
    current_user: CurrentUser,
    admin_session: AdminSessionDep,
) -> PotResponse:
    return await pot_service.create_pot(
        admin_session,
        body.group_id,
        current_user,
        base_currency=body.base_currency,
        name=body.name,
        visibility=body.visibility,
    )


# Updates a pot's name or visibility. Group admin only (403 otherwise). Only provided fields change.
@router.put("/{pot_id}", response_model=PotResponse)
async def update_pot(
    pot_id: int,
    body: PotUpdate,
    current_user: CurrentUser,
    session: SessionDep,
) -> PotResponse:
    return await pot_service.update_pot(session, pot_id, current_user, name=body.name, visibility=body.visibility)


# Deletes a pot. Group admin only; refused with 409 while it still holds anything.
@router.delete("/{pot_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_pot(pot_id: int, current_user: CurrentUser, session: SessionDep) -> None:
    await pot_service.delete_pot(session, pot_id, current_user)


# Grants or changes one member's access to a pot. Group admin only — and being an admin is what lets
# you SET this, never what lets you read anything.
@router.put("/{pot_id}/permissions/{member_id}", response_model=PotResponse)
async def set_permission(
    pot_id: int,
    member_id: int,
    body: PotPermissionUpdate,
    current_user: CurrentUser,
    session: SessionDep,
) -> PotResponse:
    return await pot_service.set_permission(session, pot_id, member_id, current_user, can_view=body.can_view, can_write=body.can_write)


# Removes a member's explicit permission row, dropping them back to the pot's visibility default.
# Group admin only.
@router.delete("/{pot_id}/permissions/{member_id}", response_model=PotResponse)
async def clear_permission(
    pot_id: int,
    member_id: int,
    current_user: CurrentUser,
    session: SessionDep,
) -> PotResponse:
    return await pot_service.clear_permission(session, pot_id, member_id, current_user)


# Lists everything the pot holds, each with its own figure and the same figure in the pot's base
# currency. Readable by whoever may see the pot at all, including a member holding 0% of it — the
# monitoring surface is not gated on owning any of it (V5).
@router.get("/{pot_id}/holdings", response_model=PotHoldingsResponse)
async def list_holdings(pot_id: int, current_user: CurrentUser, session: SessionDep) -> PotHoldingsResponse:
    return await pot_service.list_holdings(session, pot_id, current_user)


# Moves holdings into the pot. Needs pot write access (403), and every named holding must be the
# caller's own private one (404 otherwise — the whole move is refused rather than partly applied).
# An account with linked entries is refused with 409: its balance derives from one user's rows, so a
# shared one would report a different figure to every member.
@router.post("/{pot_id}/holdings", response_model=PotResponse)
async def add_holdings(
    pot_id: int,
    body: PotHoldingsMove,
    current_user: CurrentUser,
    session: SessionDep,
) -> PotResponse:
    return await pot_service.move_holdings(session, pot_id, current_user, investment_ids=body.investment_ids, account_ids=body.account_ids, into=True)


# Moves holdings back out of the pot into the caller's own private scope. Needs pot write access, and
# every named holding must already belong to THIS pot.
@router.post("/{pot_id}/holdings/remove", response_model=PotResponse)
async def remove_holdings(
    pot_id: int,
    body: PotHoldingsMove,
    current_user: CurrentUser,
    session: SessionDep,
) -> PotResponse:
    return await pot_service.move_holdings(
        session, pot_id, current_user, investment_ids=body.investment_ids, account_ids=body.account_ids, into=False
    )


# Lists the pot's ownership ledger in replay order. Visible to whoever may see the pot at all,
# including a member holding 0% of it.
@router.get("/{pot_id}/ownership", response_model=list[PotOwnershipEventResponse])
async def list_ownership_events(pot_id: int, current_user: CurrentUser, session: SessionDep) -> list[PotOwnershipEventResponse]:
    return await pot_ownership_service.list_events(session, pot_id, current_user)


# Records the pot's opening baseline: its value and each owner's percentage on a date. Refused with
# 409 once a baseline exists, and with 400 when the percentages do not total 100.
# Returns one event per owner — the baseline is a division, not a single movement.
@router.post("/{pot_id}/ownership/opening", response_model=list[PotOwnershipEventResponse], status_code=status.HTTP_201_CREATED)
async def record_opening(
    pot_id: int,
    body: PotOpeningCreate,
    current_user: CurrentUser,
    session: SessionDep,
) -> list[PotOwnershipEventResponse]:
    return await pot_ownership_service.record_opening(
        session,
        pot_id,
        current_user,
        date=body.date,
        value=body.value,
        shares=body.shares,
        notes=body.notes,
    )


# Records a contribution or a withdrawal: money crossing the scope boundary, priced at the pot's unit
# price on the date. Needs pot write access (403 otherwise) and a known valuation on that date (400).
@router.post("/{pot_id}/ownership/movements", response_model=PotOwnershipEventResponse, status_code=status.HTTP_201_CREATED)
async def record_movement(
    pot_id: int,
    body: PotMovementCreate,
    current_user: CurrentUser,
    session: SessionDep,
) -> PotOwnershipEventResponse:
    return await pot_ownership_service.record_movement(
        session,
        pot_id,
        current_user,
        type=body.type,
        date=body.date,
        member_id=body.member_id,
        amount=body.amount,
        amount_currency=body.amount_currency,
        base_amount=body.base_amount,
        from_account_id=body.from_account_id,
        to_account_id=body.to_account_id,
        notes=body.notes,
    )


# Records a re-agreement: units moving between two members with no money at all. Taken as a
# percentage of the whole pot, because percentages go in and percentages come out.
@router.post("/{pot_id}/ownership/reagreements", response_model=PotOwnershipEventResponse, status_code=status.HTTP_201_CREATED)
async def record_reagreement(
    pot_id: int,
    body: PotReagreementCreate,
    current_user: CurrentUser,
    session: SessionDep,
) -> PotOwnershipEventResponse:
    return await pot_ownership_service.record_reagreement(
        session,
        pot_id,
        current_user,
        date=body.date,
        from_member_id=body.from_member_id,
        to_member_id=body.to_member_id,
        percentage=body.percentage,
        notes=body.notes,
    )


# Deletes an ownership event. Balances are derived, so the series simply recomputes without it.
@router.delete("/{pot_id}/ownership/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_ownership_event(pot_id: int, event_id: int, current_user: CurrentUser, session: SessionDep) -> None:
    await pot_ownership_service.delete_event(session, pot_id, event_id, current_user)
